File: PlayUnfair.py
#Python Playfair cipher algorithm

import pdb
import math

#alphabet = [a-z]
alphabet = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "u", "v", "w", "x", "y"]
alphabob = []

def phormat(txt) :
	txt = txt.replace(" ", "")
	txt = txt.lower()
	txt	= txt.replace("z", "x")
	txt = splt(txt)
	return txt

def splt(word) :
	return [char for char in word]

#function to create the coded index
def indexus(key) :
	key.extend(alphabet)
	for x in key :
		if x not in alphabob :
			alphabob.append(x)
	return alphabob

#function to encrypt message with given index
def encrypt(plaintext, idx) :
	#something funky
	if len(plaintext) % 2 == 1 :
		plaintext.append("x")
	ciphertext = ""
	enutext = enumerate(plaintext)
	for pos,letr in enutext :
		idx = alphabob.index(letr)	
		if pos % 2 == 0 :
			posPlus = alphabob.index(plaintext[pos + 1])
			edx = 5 * (math.floor(idx / 5)) + posPlus - 5 * (math.floor(posPlus / 5))
			if edx > 24 :
				edx = edx - 24
			if edx < 0 :
				edx = edx + 24
			edx = int(edx)
			echar = alphabob[edx]
			if echar == plaintext[pos] :
				echar = plaintext[pos + 1]
			ciphertext += echar
		else :
			posNeg = alphabob.index(plaintext[pos - 1])
			edx = 5 * (math.floor(idx / 5)) + posNeg - 5 * (math.floor(posNeg / 5))
			if edx > 24 :
				edx = edx - 24
			if edx < 0 :
				edx = edx + 24
			edx = int(edx)
			echar = alphabob[edx]
			if echar == plaintext[pos] :
				echar = plaintext[pos - 1]
			ciphertext += echar
	print ciphertext
	return ciphertext

key = raw_input("Enter a codeword: ")
key = phormat(key)

plaintext = raw_input("Enter some text to encrypt or decrypt: ")
plaintext = phormat(plaintext)

butt = indexus(key)
encrypt(plaintext, butt)

#encrypt(plaintext, indexus(key))

File: PlayUnfair.pl
#!/usr/bin/perl
use strict;
use warnings;
use List::MoreUtils qw(uniq first_index);
use POSIX qw(floor);

#Perl Playfair Cipher Algorithm

my @alphabet = qw(a b c d e f g h i j k l m n o p q r s t u v w x y);

print "enter a codeword: ";
my $codeword = <STDIN>;
chomp $codeword;
$codeword =~ tr/A-Z/a-z/;
#removes whitespace
$codeword =~ s/\s+//g;
$codeword =~ s/z/x/g;

my @key = split //, $codeword;
#adds key to beginning of alphabet
unshift (@alphabet, @key);
#removes repeated characters from key alphabet
my @indeks = uniq(@alphabet);

print "enter a message: ";
my $message = <STDIN>;
chomp $message;
#removes whitespace
$message =~ s/\s+//g;
$message =~ s/z/x/g;
my @plaintext = split //, $message;

if (@plaintext % 2 == 1)
{
    push(@plaintext, "x");
}

my $ciphertext;

#$pos gets numbers 0-$#plaintext
for my $pos (0..$#plaintext)
{
    #letter at each $pos
    my $letr = $plaintext[$pos];
    #index of each letter in @indeks
    my $idx = first_index {$_ eq $plaintext[$pos]} @indeks;
    my $eletr;

    if ($pos % 2 == 0)
    {
        my $posPlus = first_index {$_ eq $plaintext[$pos + 1]} @indeks;
        #new encrypted position of letter in indeks
        my $edx = 5 * (floor($idx / 5)) + $posPlus - 5 * (floor($posPlus / 5));
        if ($edx > 24)
        {
            $edx = $edx - 24;
        }
        if ($edx < 0)
        {
            $edx = $edx + 24;
        }
        $eletr = $indeks[$edx];
        if ($eletr eq $plaintext[$pos])
        {
            $eletr = $plaintext[$pos + 1]
        }
        $ciphertext .= $eletr;
    }
    else
    {
        my $posNeg = first_index {$_ eq $plaintext[$pos - 1]} @indeks;
        my $edx = 5 * (floor($idx / 5)) + $posNeg - 5 * (floor($posNeg / 5));
        if ($edx > 24)
        {
            $edx = $edx - 24;
        }
        if ($edx < 0)
        {
            $edx = $edx + 24;
        }
        $eletr = $indeks[$edx];
        if ($eletr eq $plaintext[$pos])
        {
            $eletr = $plaintext[$pos - 1]
        }
        $ciphertext .= $eletr;
    }
}
print "$ciphertext\n";